test/e2e/xfn_test.go

```diff
@@ -0,0 +1,207 @@
+						return false, nil //nolint:nilerr // we want to retry and to throw error
+					}
+					return true, nil
+				}, wait.WithTimeout(1*time.Minute))
+				if err != nil {
+					t.Fatal("copying image to registry not successful", err)
+				}
+				return ctx
+			}).
+			WithSetup("CrossplaneDeployedWithFunctionsEnabled", funcs.AllOf(
+				funcs.AsFeaturesFunc(funcs.HelmUpgrade(
+					HelmOptions(
+						helm.WithArgs(
+							"--set args={--debug,--enable-composition-functions}",
+							"--set xfn.enabled=true",
+							"--set xfn.args={--debug}",
+							"--set registryCaBundleConfig.name=reg-ca",
+							"--set registryCaBundleConfig.key=domain.crt",
+							"--set xfn.resources.requests.cpu=100m",
+							"--set xfn.resources.limits.cpu=100m",
+						),
+						helm.WithWait())...)),
+				funcs.ReadyToTestWithin(1*time.Minute, namespace),
+			)).
+			WithSetup("ProviderNopDeployed", funcs.AllOf(
+				funcs.ApplyResources(FieldManager, manifests, "prerequisites/provider.yaml"),
+				funcs.ApplyResources(FieldManager, manifests, "prerequisites/definition.yaml"),
+				funcs.ResourcesCreatedWithin(30*time.Second, manifests, "prerequisites/provider.yaml"),
+				funcs.ResourcesCreatedWithin(30*time.Second, manifests, "prerequisites/definition.yaml"),
+				funcs.ResourcesHaveConditionWithin(1*time.Minute, manifests, "prerequisites/definition.yaml", v1.WatchingComposite()),
+			)).
+			Assess("CompositionWithFunctionIsCreated", funcs.AllOf(
+				funcs.ApplyResources(FieldManager, manifests, "composition.yaml"),
+				funcs.ResourcesCreatedWithin(30*time.Second, manifests, "composition.yaml"),
+			)).
+			Assess("ClaimIsCreated", funcs.AllOf(
+				funcs.ApplyResources(FieldManager, manifests, "claim.yaml"),
+				funcs.ResourcesCreatedWithin(30*time.Second, manifests, "claim.yaml"),
+			)).
+			Assess("ClaimBecomesAvailable", funcs.ResourcesHaveConditionWithin(5*time.Minute, manifests, "claim.yaml", xpv1.Available())).
+			Assess("ManagedResourcesProcessedByFunction", func(ctx context.Context, t *testing.T, config *envconf.Config) context.Context {
+				labelName := "labelizer.xfn.crossplane.io/processed"
+				rg := utils.NewResourceGetter(ctx, t, config)
+				claim := rg.Get("fn-labelizer", "default", "nop.example.org/v1alpha1", "NopResource")
+				r := utils.ResourceValue(t, claim, "spec", "resourceRef")
+
+				xr := rg.Get(r["name"], "default", r["apiVersion"], r["kind"])
+				mrefs := utils.ResourceSliceValue(t, xr, "spec", "resourceRefs")
+				for _, mref := range mrefs {
+					err := wait.For(func() (done bool, err error) {
+						mr := rg.Get(mref["name"], "default", mref["apiVersion"], mref["kind"])
+						l, found := mr.GetLabels()[labelName]
+						if !found {
+							return false, nil
+						}
+						if l != "true" {
+							return false, nil
+						}
+						return true, nil
+					}, wait.WithTimeout(5*time.Minute))
+					if err != nil {
+						t.Fatalf("Expected label %v value to be true", labelName)
+					}
+
+				}
+				return ctx
+			}).
```

Comment:
@pedjak, just noticed it's actually missing the teardown at the end, can you handle that before getting it merged?

Reply:
added